File: stroom-auth-svc/src/integration-test/java/stroom/auth/service/resources/UserResource_read_IT.java
/*
 * Copyright 2017 Crown Copyright
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package stroom.auth.service.resources;

import com.mashape.unirest.http.exceptions.UnirestException;
import org.jose4j.lang.JoseException;
import org.junit.Test;
import stroom.auth.AuthenticationFlowHelper;
import stroom.auth.resources.user.v1.User;
import stroom.auth.service.ApiException;
import stroom.auth.service.ApiResponse;
import stroom.auth.service.api.UserApi;
import stroom.auth.service.resources.support.Base_IT;

import java.io.IOException;
import java.util.List;
import java.util.UUID;

import static org.assertj.core.api.Assertions.assertThat;
import static org.assertj.core.api.Java6Assertions.fail;

public final class UserResource_read_IT extends Base_IT {

    @Test
    public final void search_users() throws UnirestException, IOException, ApiException, JoseException {
        UserApi userApi = SwaggerHelper.newUserApiClient(AuthenticationFlowHelper.authenticateAsAdmin());
        ApiResponse<String> response = userApi.getAllWithHttpInfo();
        assertThat(response.getStatusCode()).isEqualTo(200);
    }

    @Test
    public final void read_current_user() throws UnirestException, IOException, InterruptedException, ApiException, JoseException {
        UserApi userApi = SwaggerHelper.newUserApiClient(AuthenticationFlowHelper.authenticateAsAdmin());
        ApiResponse<String> response = userApi.readCurrentUserWithHttpInfo();
        List<User> user = userManager.deserialiseUsers(response.getData());

        if (user != null) {
            assertThat(user.get(0).getEmail()).isEqualTo("admin");
        } else fail("No users found");
        assertThat(response.getStatusCode()).isEqualTo(200);
    }

    @Test
    public final void read_user_that_doesnt_exist() throws UnirestException, ApiException, JoseException {
        UserApi userApi = SwaggerHelper.newUserApiClient(AuthenticationFlowHelper.authenticateAsAdmin());
        try {
            userApi.getUser(129387298);
            fail("Expected a 404 exception!");
        }catch(ApiException e) {
            assertThat(e.getCode()).isEqualTo(404);
        }
    }

    @Test
    public final void read_other_user_with_authorisation() throws UnirestException, ApiException, JoseException {
        UserApi userApi = SwaggerHelper.newUserApiClient(AuthenticationFlowHelper.authenticateAsAdmin());

        ApiResponse<Integer> response = userApi.createUserWithHttpInfo(new stroom.auth.service.api.model.User()
            .email("read_other_user_with_authorisation_" + UUID.randomUUID().toString())
            .password("password"));
        assertThat(response.getStatusCode()).isEqualTo(200);
        assertThat(response.getData()).isNotNull();

        String user = userApi.getUser(response.getData());
        assertThat(response.getStatusCode()).isEqualTo(200);
        assertThat(response.getData()).isNotNull();
    }

    @Test
    public final void read_other_user_without_authorisation() throws UnirestException, ApiException, JoseException {
        UserApi adminUserApi = SwaggerHelper.newUserApiClient(AuthenticationFlowHelper.authenticateAsAdmin());

        String userEmailA = "userEmailA_" + UUID.randomUUID().toString();
        ApiResponse<Integer> responseA = adminUserApi.createUserWithHttpInfo(new stroom.auth.service.api.model.User()
                .email(userEmailA)
                .password("password"));

        String userEmailB = "userEmailB_" + UUID.randomUUID().toString();
        ApiResponse<Integer> responseB = adminUserApi.createUserWithHttpInfo(new stroom.auth.service.api.model.User()
                .email(userEmailB)
                .password("password"));

        UserApi userApiA = SwaggerHelper.newUserApiClient(AuthenticationFlowHelper.authenticateAs(userEmailA, "password"));
        try {
            userApiA.getUser(responseB.getData());
            fail("Expected a 401!");
        } catch(ApiException e){
            assertThat(e.getCode()).isEqualTo(401);
        }
    }

}

File: stroom-auth-svc/src/integration-test/java/stroom/auth/service/resources/UserResource_read_IT.kt
/*
 * Copyright 2017 Crown Copyright
 *
 * This file is part of Stroom-Stats.
 *
 * Stroom-Stats is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * Stroom-Stats is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with Stroom-Stats.  If not, see <http://www.gnu.org/licenses/>.
 */

package stroom.auth.service.resources

import com.github.kittinunf.fuel.httpGet
import org.assertj.core.api.Assertions
import org.assertj.core.api.Assertions.assertThat
import org.assertj.core.api.Assertions.fail
import org.junit.Test


class UserResource_read_IT: UserResource_IT() {

    @Test
    fun search_users() {
        val jwsToken = login()
        val url = ROOT_URL + "?fromEmail=&usersPerPage=10&orderBy=email"
        val (_, response) = url
                .httpGet()
                .header(jsonContentType("Authorization" to "Bearer $jwsToken"))
                .responseString()
        val userJson = String(response.data)
        val users = userListMapper().fromJson(userJson)
        assertThat(response.httpStatusCode).isEqualTo(200)
    }

    @Test
    fun read_current_user() {
        val jwsToken = login()
        val (_, response) = ME_URL
                .httpGet()
                .header(jsonContentType("Authorization" to "Bearer $jwsToken"))
                .responseString()
        assertThat(response.httpStatusCode).isEqualTo(200)
        val userJson = String(response.data)

        val user = userListMapper().fromJson(userJson)

        if(user != null) {
            assertThat(user[0].email).isEqualTo("admin@")
        }
        else fail("No users found")
    }

    @Test
    fun read_user_that_doesnt_exist() {
        val jwsToken = login()
        val url = ROOT_URL + "97862345983458"
        val (_, response) = url
                .httpGet()
                .header(jsonContentType("Authorization" to "Bearer $jwsToken"))
                .responseString()
        assertThat(response.httpStatusCode).isEqualTo(404)
    }
}